### PR TITLE
EPMLSTRCMW-200 fix: Stop application in case of initialization error

### PR DESCRIPTION
**Pull request by gominasian**
Application stops when initialization error is met
cromwell.pipeline excluded from coverage

### CODE REVIEW

**Comment by OlegKoshelev** [comment]
Add a description to your task, please

**Review by DruzhininPavel — APPROVED** [review]
No questions.

File: build.sbt [hunk 1]
```diff
@@ -28,7 +28,8 @@ lazy val commonSettings = Seq(
   testOptions in Test += Tests.Argument(TestFrameworks.ScalaTest, "-h", "target/test-reports"),
   coverageEnabled in Test := true,
   coverageMinimum := 60,
-  coverageFailOnMinimum := true
+  coverageFailOnMinimum := true,
+  coverageExcludedPackages := "CromwellPipelineApp\\..*"
```
Comment (Nikolai-Isaev) [inline build.sbt]: Maybe we should put more entities here :)

**Review by Nikolai-Isaev — APPROVED** [review]

File: build.sbt (hunk 1 above)
Comment (myazinn) [reply build.sbt]: Definitely. We should exclude from coverage everything in `portal` module.

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 2]
```diff
@@ -37,16 +38,26 @@ object CromwellPipelineApp extends App {
...
-    routeCombiner(
-      userController.route,
-      projectController.route,
-      projectFileController.route,
-      runController.route,
-      configurationController.route
-    )
-  }
+  try {
+    val route = authController.route ~ securityDirective.authenticated {
```
Comment (myazinn) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: By this moment application could have faced tons of other errors. Maybe we couldn't create some service, or maybe database migration failed.
Basically every step that comes after ActorSystem creation must be treated as unsafe step, i.e. included in try-catch statement.

**Review by myazinn — CHANGES_REQUESTED** [review]

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala (hunk 2 above)
Comment (myazinn) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: @gominasian probably you didn't notice it

File: build.sbt [hunk 3]
```diff
@@ -28,7 +28,8 @@ lazy val commonSettings = Seq(
   testOptions in Test += Tests.Argument(TestFrameworks.ScalaTest, "-h", "target/test-reports"),
   coverageEnabled in Test := true,
   coverageMinimum := 60,
-  coverageFailOnMinimum := true
+  coverageFailOnMinimum := true,
+  coverageExcludedPackages := "<empty>;cromwell\\.pipeline.*"
```
Comment (myazinn) [inline build.sbt]: Something is wrong. I don't see a lot of module here
https://coveralls.io/builds/39858444
Compare this to another report
https://coveralls.io/builds/39810273

**Review by myazinn — CHANGES_REQUESTED** [review]

**Review by DruzhininPavel — APPROVED** [review]
Ok. Now it is work.

**Comment by iliaonishchenko** [comment]
Is changes request fixed?
Can I review it? MR is in draft status.

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 4]
```diff
@@ -10,43 +10,58 @@ import cromwell.pipeline.datastorage.dto.auth.AccessTokenContent
...
+
   implicit val system: ActorSystem = ActorSystem("cromwell-pipeline")
   implicit val executionContext: ExecutionContextExecutor = system.dispatcher
 
-  val log = LoggerFactory.getLogger(CromwellPipelineApp.getClass)
   val components = new ApplicationComponents()
 
   import components.applicationConfig.webServiceConfig
   import components.controllerModule._
   import components.datastorageModule._
```
Comment (iliaonishchenko) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Why not to move these lines inside `Try`?

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 5]
```diff
@@ -10,43 +10,58 @@ import cromwell.pipeline.datastorage.dto.auth.AccessTokenContent
...
-
-  val route = authController.route ~ securityDirective.authenticated {
-    routeCombiner(
-      userController.route,
-      projectController.route,
-      projectFileController.route,
-      runController.route,
-      configurationController.route
-    )
+  val isSuccessful = Try {
```
Comment (iliaonishchenko) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Should we have such value at all?
We can do matching at `Try` statement.

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 6]
```diff
@@ -10,43 +10,58 @@ import cromwell.pipeline.datastorage.dto.auth.AccessTokenContent
...
   }
 
-  log.info(s"Server online at http://${webServiceConfig.interface}:${webServiceConfig.port}/")
-  Http().newServerAt(webServiceConfig.interface, webServiceConfig.port).bindFlow(route)
+  isSuccessful match {
+    case Success(_) =>
+      log.info(s"Application is running")
+    case Failure(e) =>
+      system.terminate()
+      log.info(s"Application cannot be started.\n${e.fillInStackTrace()}")
```
Comment (iliaonishchenko) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: When you log error, it's probably better to use `error` level.

**Review by iliaonishchenko — CHANGES_REQUESTED** [review]
Nice job, there're some comments:

**Review by OlegKoshelev — APPROVED** [review]

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala (hunk 5 above)
Comment (gominasian) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Thank you, will fix it

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala (hunk 6 above)
Comment (gominasian) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Fixed it

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala (hunk 5 above)
Comment (gominasian) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Done

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala (hunk 4 above)
Comment (gominasian) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Yep, fixed

**Review by GrigoriiBerezin — APPROVED** [review]
Nice

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 7]
```diff
@@ -10,43 +10,56 @@ import cromwell.pipeline.datastorage.dto.auth.AccessTokenContent
...
+
+    val route = Route.seal(authController.route ~ securityDirective.authenticated {
+      routeCombiner(
+        userController.route,
+        projectController.route,
+        projectFileController.route,
+        runController.route,
+        configurationController.route
+      )
+    })(rejectionHandler = rejectionHandler)
```
Comment (Dmitrii-Shevchenko) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Why do you signify rejectionHandler explicitly?

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 8]
```diff
@@ -10,43 +10,56 @@ import cromwell.pipeline.datastorage.dto.auth.AccessTokenContent
...
+    implicit val rejectionHandler: RejectionHandler = RejectionHandler
+      .newBuilder()
+      .handle {
+        case ValidationRejection(msg, _)      => complete(StatusCodes.BadRequest, msg)
+        case MissingAccessTokenRejection(msg) => complete(StatusCodes.Unauthorized, msg)
+      }
+      .result()
+      .withFallback(RejectionHandler.default)
+
+    val route = Route.seal(authController.route ~ securityDirective.authenticated {
```
Comment (Dmitrii-Shevchenko) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: What is the reason that you add routes via seal?

**Comment by Dmitrii-Shevchenko** [comment]
Squash commits please

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 9]
```diff
@@ -10,43 +10,56 @@ import cromwell.pipeline.datastorage.dto.auth.AccessTokenContent
...
+
+    log.info(s"Server online at http://${webServiceConfig.interface}:${webServiceConfig.port}/")
+    Http().newServerAt(webServiceConfig.interface, webServiceConfig.port).bindFlow(route)
+
+  } match {
+    case Success(_) =>
+      log.info(s"Application is running")
+    case Failure(e) =>
+      system.terminate()
+      log.error(s"Application cannot be started.\n${e.fillInStackTrace()}")
```
Comment (iliaonishchenko) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: When you log at `error` level, you provide text description and an error as a second argument. No need to interpolate Exception to string.

File: build.sbt [hunk 10]
```diff
@@ -29,7 +29,7 @@ lazy val commonSettings = Seq(
   coverageEnabled in Test := true,
   coverageMinimum := 60,
   coverageFailOnMinimum := true,
-  coverageExcludedPackages := "cromwell.pipeline.database.*"
+  coverageExcludedPackages := "<empty>; cromwell\\.pipeline; cromwell\\.pipeline\\.database.*; "
```
Comment (iliaonishchenko) [inline build.sbt]: Could you please describe what it does?

Comment (iliaonishchenko) [reply build.sbt]: Why there is `empty` in front of the string?

Comment (gominasian) [reply build.sbt]: This are regex to exclude some files from coverage:
I used escape slashes (\\) to make dots match literally.

- empty tag uses for pick up first exclusion. To be honest, it works perfect without it too, but I think there are cases when you really need this tag. And it's in examples in scoverage docs but without any description.

- cromwell\\.pipeline excludes files straight from cromwell.pipeline, but not from cromwell.pipeline.database, for example

- cromwell\\.pipeline\\.database.* excludes everything from cromwell.pipeline.database and inside packages (but it is from some other commit, I just put escape slashes there)

**Comment by iliaonishchenko** [comment]
Build fails.

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 11]
```diff
@@ -10,24 +10,14 @@ import cromwell.pipeline.datastorage.dto.auth.AccessTokenContent
...
 
 import scala.concurrent.ExecutionContextExecutor
+import scala.util.{ Failure, Success, Try }
 
 object CromwellPipelineApp extends App {
 
-  implicit val system: ActorSystem = ActorSystem("cromwell-pipeline")
-  implicit val executionContext: ExecutionContextExecutor = system.dispatcher
-
   val log = LoggerFactory.getLogger(CromwellPipelineApp.getClass)
```
Comment (Dmitrii-Shevchenko) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala:18]: It would be better if you keep the previous order

Comment (gominasian) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala:18]: Which one?

Comment (Dmitrii-Shevchenko) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala:18]: For example variable `log` and implicites.
You replaced the lines in order, but that doesn't make any sense.

Comment (gominasian) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala:18]: Is it important to keep previous order? I've put log before implicits, bc I want them to stay together with rejectionHandler, but it's a big block of code and logger could be inconspicuous this way

**Review by Dmitrii-Shevchenko — APPROVED** [review]
Good work!

**Review by iliaonishchenko — CHANGES_REQUESTED** [review]
Nice work! I'm ok with all changes, the only thing you need to do is to change commit message according to the project's guidelines.

**Review by iliaonishchenko — APPROVED** [review]
Looks good to me!

File: project/Dependencies.scala [hunk 12]
```diff
@@ -58,6 +59,7 @@ object Dependencies {
   val logback = "ch.qos.logback" % "logback-classic" % Version.logback
   val wireMock = "com.github.tomakehurst" % "wiremock" % Version.wireMock % Test
   val sl4j = "org.slf4j" % "slf4j-api" % Version.sl4j
+  val slf4j = "org.slf4j" % "slf4j-api" % Version.slf4j
```
Comment (myazinn) [inline project/Dependencies.scala]: Why do we need the same dependency twice?

**Review by myazinn — COMMENTED** [review]
Main changes look good to me. Although I have a question about new dependency.

File: project/Dependencies.scala (hunk 12 above)
Comment (gominasian) [reply project/Dependencies.scala]: Fix it

**Review by myazinn — APPROVED** [review]

**Pull request merged**